core/types/calls.cc

```diff
@@ -1916,6 +1916,45 @@ class Array_flatten : public IntrinsicMethod {
     }
 } Array_flatten;
 
+class Array_product : public IntrinsicMethod {
+public:
+    void apply(Context ctx, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        // Unwrap the array one time to get the element type (we'll rewrap it down at the bottom)
+        TypePtr element;
+        if (auto *ap = cast_type<AppliedType>(thisType)) {
+            ENFORCE(ap->klass == Symbols::Array() || ap->klass.data(ctx)->derivesFrom(ctx, Symbols::Array()));
+            ENFORCE(!ap->targs.empty());
+            element = ap->targs.front();
+        } else if (auto *tuple = cast_type<TupleType>(thisType)) {
+            element = tuple->elementType();
+        } else {
+            ENFORCE(false, "Array#product on unexpected type: {}", args.selfType->show(ctx));
+        }
+
+        vector<TypePtr> unwrappedElems;
+        unwrappedElems.reserve(args.args.size() + 1);
+        unwrappedElems.emplace_back(element);
+
+        for (auto arg : args.args) {
+            auto argTyp = arg->type;
+            TypePtr el;
+            if (auto *ap = cast_type<AppliedType>(argTyp.get())) {
+                ENFORCE(ap->klass == Symbols::Array() || ap->klass.data(ctx)->derivesFrom(ctx, Symbols::Array()));
+                ENFORCE(!ap->targs.empty());
+                el = ap->targs.front();
+            } else if (auto *tuple = cast_type<TupleType>(argTyp.get())) {
+                el = tuple->elementType();
+            } else {
+                ENFORCE(false, "Array#product called with unexpected type: {}", el->show(ctx));
```

Comment:
ditto, but you can have _any_ type here in practice.

Reply:
Take a look at `beginError` from the `Array#flatten` case to learn how it works. You probably want to either invent a new error code, or re-use the one that would normally result from mismatching an argument type:

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams(x%3A%20Integer).void%7D%0A%20%20def%20self.foo(x)%3B%20end%0Aend%0A%0Adef%20main%0A%20%20A.foo(%22%22)%0Aend

^ "Expected Integer but found String" is error code 7002 which maps to `MethodArgumentMismatch`.

Reply:
Thanks, fixed!